src/crimson/os/seastore/cache.cc

```diff
@@ -976,6 +979,7 @@ record_t Cache::prepare_record(Transaction &t)
 	  final_crc,
 	  (segment_off_t)i->get_length(),
 	  i->get_version() - 1,
+	  i->last_modified.time_since_epoch().count(),
```

Comment:
Why duplicate this for each delta instead of writing it once on the whole record?

Reply:
Ah, yes, will correct it.

Reply:
or probably to the whole `record_group_header_t`

Reply:
> or probably to the whole `record_group_header_t`

I think, to use the time that the record group is formed and persisted to journal as the last modification time, we'll probably have to pass a callback into the journal when submitting the record, as the record group is hidden from the transaction manager.

Is this acceptable?

Reply:
IIUC we already mark the rewrite extents as delay and will always write them to the ool record in a separate segment, right? So I think we don't need to worry about the last modification time when submitting journal records -- just always now().

When scanning records, I think the last modification time from `record_group_header_t` is available to the scanner.

Does it make sense?

Reply:
> IIUC we already mark the rewrite extents as delay and will always write them to the ool record in a non-journal(ool) segment, right? So I think we don't need to worry about the last modification time when submitting journal records -- just always now().

I have to not only set the time of commit to record_group_header_t, but also all extents' last_modified and last_rewritten, right now, record group commit is within the journal, so I think we have to pass a callback into the journal, so that we can set `CachedExtent::last_modified` and `CachedExtent::last_rewritten` to the time of record group commit.

Reply:
last_modified and last_rewritten are only valid for the rewrite extents. They (except for lba extents) seems to me are always written to the ool records which are written by the Writer, not Journal, is that correct?

Reply:
> IIUC we already mark the rewrite extents as delay and will always write them to the ool record in a non-journal(ool) segment, right?

I don't think so, lba extents are always inline, and logical extents are always delayed alloc'd but not always ool

Reply:
> They (except for lba extents) seems to me are always written to the ool records which are written by the Writer, not Journal, is that correct?

Yes, but I don't think it matters, all kinds of extents need to be taken into account, and besides last_rewritten, we have to set last_modified.

Reply:
Seems to me currently delayed extents are always ool. Even if it is not, we can make sure to rewrite the logical extents to ool records in non-journal segments, so they will not be mixed with freshly committed inline extents physically.

Reply:
> Seems to me currently delayed extents are always ool. Even if it is not, we can make sure to rewrite the logical extents to ool records in non-journal segments, so they will not be mixed with freshly committed inline extents physically.

I don't quite follow, what's your point? Even if we separate rewritten logical extents with inline extents, we still have to set lba extents' last_rewritten and all extents' last_modified, right? And if we choose to use the time of record group commit, we still have to consider the journal, right?

Reply:
Maybe CachedExtent doesn't have to keep the last_modified and last_rewritten time, right now, all we care about is the last_modified time of the segment.

But this approach diverges from the spritlfs which takes the "youngest extent"'s modification time as an import to calculate the age. Altough, now, we are not taking extents into account when calculating the segments' age either, at least, we keep tracking their modification time.

@athanatos What do you think?

Reply:
> And what we are discussing here is the time of the modification of extents represented by deltas which only exists in the journal records.

I think we don't need to worry about the last modification time of extents/deltas in journal records, they are always now() and can be stored in `record_group_header_t`. Maybe I'm still missing something.

Reply:
> I think we don't need to worry about the last modification time of extents/deltas in journal records, they are always now() and can be stored in `record_group_header_t`. Maybe I'm still missing something.

I think we need to either consider the last modification and last rewritten time of extents/deltas or not consider them at all, I don't think it makes sense to only consider the last rewritten time, because I think both have impact on the extents' age.

But maybe we don't have to care about extents' age for now, maybe only segment age is enough, which is what I'm suggesting [here](https://github.com/ceph/ceph/pull/43474#discussion_r788529552).

Reply:
After a second thought, I think we may still need to keep track of the modification/rewritten time of the extents. Because our final purpose is to put extents with similar age in the same segment, so we can't omit that age. Am I right?

Reply:
I think to put extents with similar age in the same segment, we have to be able to reclaim multiple segments and sort all live extents in those segments after which we rewrite those live extents. Should I implement it in this PR, or a follow up one?

@athanatos @cyx1231st

Reply:
> > And what we are discussing here is the time of the modification of extents represented by deltas which only exists in the journal records.
> 
> I think we don't need to worry about the last modification time of extents/deltas in journal records, they are always now() and can be stored in `record_group_header_t`. Maybe I'm still missing something.

I think we have some misunderstanding here, I meant we have to set the last modification time and last rewritten time of all related in-cache extents.

Reply:
TBH I'm not sure what is the main subject of this thread now. I feel distracted. IIUC there are at least the following different topics:

1. Should we track the age in the granularity of segment, record, or extent?
   * TBH I don't have a definate answer right now, but I think the record granularity might work by maintaining an average age of the extents within the same record. It is more accurate and flexible than the segment-level age, and don't need the effort to persist the age for every extent.
2. Should we track the age of in-cache extents?
   * I think we only need to track the age of dirty extents in cache, so we can identify the average age when rewriting dirty extents into ool records.
3. If we track the age in the granularity of record, where should we store it?
   * I think `record_group_header_t` might be a better place.
4. If we store the age in `record_group_header_t`, should we pass a callback into the journal when submitting the journal record?
   * I think for journal records, the age to submit is always now(), so no need to pass any callback.

Reply:
I think our issues now are as follows:

1. should we take the time of record group commit as the modification time?
   pros: we don't have to store modification time in every record header, which is more efficient.
   problem: I think we'll have to set the modification time of all in-cache extents that are related to the transaction and the written segment to the time of record group commit, so we probably will have to pass a callback into the journal, because record group commit is not visible outside the journal; otherwise, we have to expose the record group commit event to other componenets like TransactionManager. Is this an acceptable pattern?

2. should we track live extents modification time or just the segments' modification time?
   First, I think we have to take extents' modification time into account when doing gc and rewrite, because we want to put extents with similar age into the same segment. The problem is should we track every extent's modification time which is the same as the record commit time or just track the segments' modification time and use it an approximation of the extents modification time when doing gc and rewrite. If we take the latter way, we can just remove `CachedExtent::last_modified` and `CachedExtent::last_rewritten`, at the cost of precision of course.

3. To put extents with similar age in the same segment, I think we have to change the segment cleaner in the following way: 1) keep scanning segments until we can rewrite at-least two segment worth of live extents; 2) sort the live extents by their age; 3) write them out. If I'm right, should I implement this mechinary in this PR, or a follow-up one?

Reply:
> I think we need to pass a callback to the journal because we have to set the last_modified/last_rewritten time of in-cache extents and the segment that's being written to the time of record group commit

We can set the "age" of the submitting mutated extents in `Cache::prepare_record` to `now`, which will be committed as dirty extents in cache.

Reply:
> We can set the "age" of the submitting mutated extents in `Cache::prepare_record` to `now`, which will be committed as dirty extents in cache.

Then we have inconsistent in-cache and ondisk modification time

Reply:
> Then we have inconsistent in-cache and ondisk modification time.

I'm not sure why does this matter... It should be within millionseconds to write down that record. And it might be days from the record submittion to record reclaim.

Reply:
> It should be within millionseconds to write down that record. And it might be days from the record submittion to record reclaim.

I think the core issue is how large the difference of the age between extents is, because the final goal is to put age-similar extents together. If that difference is over 10 ms, then maybe we can omit this inconsistency, otherwise we'd better not do this. I think this is up to the user IO pattern, when SeaStore is put in use in online clusters, the iops could 100K+, which means over 100 IOs in one millisecond. I think we'd better not make any assumption of the impact of this consistency for now, at least not for just moving modification time from record header to record group header.

Reply:
JournalSubmitter is designed to submit records as soon as possible, and group the pending records when possible, it should be rare that JournalSubmitter is saturized to block the submitter.

However, I don't think it is a big problem whether the age is in `record_group_header_t` or `record_header_t`.

Anyway, I don't think it worth the complexity to introduce a callback into Journal for this reason.